File: Python/Subsets.py
class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
    def subsets(self, S):
        if not S:
            return []
        if len(S) == 1:
            return [[],S]
        
        pre = self.subsets(S[:-1])
        ret = pre[:]
#        ret.append([S[-1]])

        for l in pre:
            newitem = l[:]
            newitem.append(S[-1])
            ret.append(sorted(newitem))
            
        return sorted(ret)

File: Java/src/Subsets.java
import java.util.ArrayList;
import java.util.Arrays;

public class Solution
{
  public List<List<Integer>> subsets(int[] S)
  {
    List<List<Integer>> ret = new ArrayList<>();
    ret.add(new ArrayList<Integer>());
    if (S == null)
      return ret;
    Arrays.sort(S);
    for (int i : S)
    {
      List<List<Integer>> cur = new ArrayList<>();
      for (List<Integer> sub : ret)
      {
        List<Integer> a = new ArrayList<>(sub);
        a.add(i);
        cur.add(a);
      }
      ret.addAll(cur);
    }
    return ret;
  }
}
